File: apps/TiflisCode/Shared/Services/WebSocket/CommandBuilder.swift
//
//  CommandBuilder.swift
//  TiflisCode
//
//  Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
//  Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
//

import Foundation

/// Factory for creating command configurations with appropriate retry and queue settings.
///
/// All commands retry on failure - no user action is silently dropped.
/// The only exception is terminal resize which uses 1 retry since only the latest size matters.
enum CommandBuilder {
    // MARK: - Supervisor Commands

    /// Create a supervisor text command
    /// - Parameters:
    ///   - command: The text command to send
    ///   - messageId: Optional message ID for tracking (defaults to new UUID)
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func supervisorCommand(_ command: String, messageId: String = UUID().uuidString) -> CommandConfig {
        let message: [String: Any] = [
            "type": "supervisor.command",
            "id": messageId,
            "payload": [
                "command": command
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.command"
        )
    }

    /// Create a supervisor voice command
    /// - Parameters:
    ///   - audioBase64: Base64-encoded audio data
    ///   - format: Audio format (e.g., "m4a")
    ///   - messageId: Message ID for tracking transcription
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func supervisorVoiceCommand(
        audioBase64: String,
        format: String,
        messageId: String
    ) -> CommandConfig {
        let message: [String: Any] = [
            "type": "supervisor.command",
            "id": UUID().uuidString,
            "payload": [
                "audio": audioBase64,
                "audio_format": format,
                "message_id": messageId
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.command.voice"
        )
    }

    /// Cancel supervisor execution
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func supervisorCancel() -> CommandConfig {
        let message: [String: Any] = [
            "type": "supervisor.cancel",
            "id": UUID().uuidString
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.cancel"
        )
    }

    /// Clear supervisor context
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func supervisorClearContext() -> CommandConfig {
        let message: [String: Any] = [
            "type": "supervisor.clear_context",
            "id": UUID().uuidString
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.clear_context"
        )
    }

    /// Create a new session
    /// - Parameters:
    ///   - sessionType: Type of session (cursor, claude, opencode, terminal)
    ///   - agentName: Optional agent alias name
    ///   - workspace: Workspace name
    ///   - project: Project name
    ///   - requestId: Request ID for tracking the response
    /// - Returns: Command configuration with 3 retries and queue enabled (critical)
    static func createSession(
        sessionType: String,
        agentName: String?,
        workspace: String,
        project: String,
        requestId: String
    ) -> CommandConfig {
        var payload: [String: Any] = [
            "session_type": sessionType,
            "workspace": workspace,
            "project": project
        ]
        if let name = agentName {
            payload["agent_name"] = name
        }

        let message: [String: Any] = [
            "type": "supervisor.create_session",
            "id": requestId,
            "payload": payload
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.create_session"
        )
    }

    /// Terminate a session
    /// - Parameter sessionId: ID of the session to terminate
    /// - Returns: Command configuration with 3 retries and queue enabled (critical)
    static func terminateSession(sessionId: String) -> CommandConfig {
        let message: [String: Any] = [
            "type": "supervisor.terminate_session",
            "id": UUID().uuidString,
            "payload": [
                "session_id": sessionId
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "supervisor.terminate_session"
        )
    }

    // MARK: - Session Commands

    /// Subscribe to a session (agent or terminal)
    /// - Parameter sessionId: ID of the session to subscribe to
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func sessionSubscribe(sessionId: String) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.subscribe",
            "session_id": sessionId
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.subscribe"
        )
    }

    /// Unsubscribe from a session
    /// - Parameter sessionId: ID of the session to unsubscribe from
    /// - Returns: Command configuration with 1 retry and no queue (non-critical)
    static func sessionUnsubscribe(sessionId: String) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.unsubscribe",
            "session_id": sessionId
        ]
        return CommandConfig(
            message: message,
            maxRetries: 1,
            shouldQueue: false,
            debugName: "session.unsubscribe"
        )
    }

    /// Execute a command in an agent session (text)
    /// - Parameters:
    ///   - sessionId: ID of the session
    ///   - content: Text content to execute
    ///   - messageId: Optional message ID for tracking (defaults to new UUID)
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func sessionExecute(sessionId: String, content: String, messageId: String = UUID().uuidString) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.execute",
            "id": messageId,
            "session_id": sessionId,
            "payload": [
                "content": content
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.execute"
        )
    }

    /// Execute a voice command in an agent session
    /// - Parameters:
    ///   - sessionId: ID of the session
    ///   - audioBase64: Base64-encoded audio data
    ///   - format: Audio format (e.g., "m4a")
    ///   - messageId: Message ID for tracking transcription
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func sessionVoiceExecute(
        sessionId: String,
        audioBase64: String,
        format: String,
        messageId: String
    ) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.execute",
            "id": UUID().uuidString,
            "session_id": sessionId,
            "payload": [
                "audio": audioBase64,
                "audio_format": format,
                "message_id": messageId
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.execute.voice"
        )
    }

    /// Cancel an agent session execution
    /// - Parameter sessionId: ID of the session
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func sessionCancel(sessionId: String) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.cancel",
            "id": UUID().uuidString,
            "session_id": sessionId
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.cancel"
        )
    }

    // MARK: - Terminal Commands

    /// Send input to a terminal session
    /// - Parameters:
    ///   - sessionId: ID of the terminal session
    ///   - data: Input data (text or raw bytes as string)
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func terminalInput(sessionId: String, data: String) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.input",
            "session_id": sessionId,
            "payload": [
                "data": data
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.input"
        )
    }

    /// Resize a terminal session
    /// - Parameters:
    ///   - sessionId: ID of the terminal session
    ///   - cols: Number of columns
    ///   - rows: Number of rows
    /// - Returns: Command configuration with 1 retry and no queue (only latest matters)
    static func terminalResize(sessionId: String, cols: Int, rows: Int) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.resize",
            "session_id": sessionId,
            "payload": [
                "cols": cols,
                "rows": rows
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 1,
            shouldQueue: false,
            debugName: "session.resize"
        )
    }

    /// Request replay of terminal history
    /// - Parameters:
    ///   - sessionId: ID of the terminal session
    ///   - sinceSequence: Sequence number to replay from (0 for all)
    ///   - limit: Maximum number of messages to retrieve
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func terminalReplay(
        sessionId: String,
        sinceSequence: Int,
        limit: Int
    ) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.replay",
            "session_id": sessionId,
            "payload": [
                "since_sequence": sinceSequence,
                "limit": limit
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.replay"
        )
    }

    /// Request replay of terminal history by timestamp
    /// - Parameters:
    ///   - sessionId: ID of the terminal session
    ///   - sinceTimestamp: Timestamp to replay from (0 for all)
    ///   - limit: Maximum number of messages to retrieve
    /// - Returns: Command configuration with 3 retries and queue enabled
    static func terminalReplayByTimestamp(
        sessionId: String,
        sinceTimestamp: Int,
        limit: Int
    ) -> CommandConfig {
        let message: [String: Any] = [
            "type": "session.replay",
            "session_id": sessionId,
            "payload": [
                "since_timestamp": sinceTimestamp,
                "limit": limit
            ]
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "session.replay"
        )
    }

    // MARK: - System Commands

    static func historyRequest(
        sessionId: String?,
        beforeSequence: Int? = nil,
        limit: Int = 20
    ) -> CommandConfig {
        var payload: [String: Any] = [:]
        if let sessionId = sessionId {
            payload["session_id"] = sessionId
        }
        if let beforeSequence = beforeSequence {
            payload["before_sequence"] = beforeSequence
        }
        payload["limit"] = limit

        let message: [String: Any] = [
            "type": "history.request",
            "id": UUID().uuidString,
            "payload": payload
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "history.request"
        )
    }

    static func sync() -> CommandConfig {
        let message: [String: Any] = [
            "type": "sync",
            "id": UUID().uuidString
        ]
        return CommandConfig(
            message: message,
            maxRetries: 3,
            shouldQueue: true,
            debugName: "sync"
        )
    }
}

File: apps/TiflisCodeAndroid/app/src/main/java/io/tiflis/code/data/websocket/CommandBuilder.kt
/*
 * Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
 * Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
 */

package io.tiflis.code.data.websocket

import java.util.UUID

/**
 * Factory for creating command configurations with appropriate retry and queue settings.
 *
 * All commands retry on failure - no user action is silently dropped.
 * The only exception is terminal resize which uses 1 retry since only the latest size matters.
 */
object CommandBuilder {

    // MARK: - Supervisor Commands

    /**
     * Create a supervisor text command.
     * @param command The text command to send
     * @param messageId Message ID for tracking
     * @return Command configuration with 3 retries and queue enabled
     */
    fun supervisorCommand(command: String, messageId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "supervisor.command")
            put("id", UUID.randomUUID().toString())
            put("payload", mapOf(
                "command" to command,
                "message_id" to messageId
            ))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.command"
        )
    }

    /**
     * Create a supervisor voice command.
     * @param audioBase64 Base64-encoded audio data
     * @param format Audio format (e.g., "m4a")
     * @param messageId Message ID for tracking transcription
     * @return Command configuration with 3 retries and queue enabled
     */
    fun supervisorVoiceCommand(
        audioBase64: String,
        format: String,
        messageId: String
    ): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "supervisor.command")
            put("id", UUID.randomUUID().toString())
            put("payload", mapOf(
                "audio" to audioBase64,
                "audio_format" to format,
                "message_id" to messageId
            ))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.command.voice"
        )
    }

    /**
     * Cancel supervisor execution.
     * @return Command configuration with 3 retries and queue enabled
     */
    fun supervisorCancel(): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "supervisor.cancel")
            put("id", UUID.randomUUID().toString())
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.cancel"
        )
    }

    /**
     * Clear supervisor context.
     * @return Command configuration with 3 retries and queue enabled
     */
    fun supervisorClearContext(): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "supervisor.clear_context")
            put("id", UUID.randomUUID().toString())
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.clear_context"
        )
    }

    /**
     * Create a new session.
     * @param sessionType Type of session (cursor, claude, opencode, terminal)
     * @param agentName Optional agent alias name
     * @param workspace Workspace name
     * @param project Project name
     * @param worktree Optional worktree name
     * @param requestId Request ID for tracking the response
     * @return Command configuration with 3 retries and queue enabled (critical)
     */
    fun createSession(
        sessionType: String,
        agentName: String?,
        workspace: String?,
        project: String?,
        worktree: String?,
        requestId: String
    ): CommandConfig {
        val payload = buildMap<String, Any?> {
            put("session_type", sessionType)
            if (agentName != null) put("agent_name", agentName)
            if (workspace != null) put("workspace", workspace)
            if (project != null) put("project", project)
            if (worktree != null) put("worktree", worktree)
        }

        val message = buildMap<String, Any?> {
            put("type", "supervisor.create_session")
            put("id", requestId)
            put("payload", payload)
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.create_session"
        )
    }

    /**
     * Terminate a session.
     * @param sessionId ID of the session to terminate
     * @return Command configuration with 3 retries and queue enabled (critical)
     */
    fun terminateSession(sessionId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "supervisor.terminate_session")
            put("id", UUID.randomUUID().toString())
            put("payload", mapOf("session_id" to sessionId))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "supervisor.terminate_session"
        )
    }

    // MARK: - Session Commands

    /**
     * Subscribe to a session (agent or terminal).
     * @param sessionId ID of the session to subscribe to
     * @return Command configuration with 3 retries and queue enabled
     */
    fun sessionSubscribe(sessionId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.subscribe")
            put("session_id", sessionId)
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.subscribe"
        )
    }

    /**
     * Unsubscribe from a session.
     * @param sessionId ID of the session to unsubscribe from
     * @return Command configuration with 1 retry and no queue (non-critical)
     */
    fun sessionUnsubscribe(sessionId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.unsubscribe")
            put("session_id", sessionId)
        }
        return CommandConfig(
            message = message,
            maxRetries = 1,
            shouldQueue = false,
            debugName = "session.unsubscribe"
        )
    }

    /**
     * Execute a command in an agent session (text).
     * @param sessionId ID of the session
     * @param text Text content to execute
     * @param messageId Message ID for tracking
     * @return Command configuration with 3 retries and queue enabled
     */
    fun sessionExecute(sessionId: String, text: String, messageId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.execute")
            put("id", UUID.randomUUID().toString())
            put("session_id", sessionId)
            put("payload", mapOf(
                "text" to text,
                "message_id" to messageId
            ))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.execute"
        )
    }

    /**
     * Execute a voice command in an agent session.
     * @param sessionId ID of the session
     * @param audioBase64 Base64-encoded audio data
     * @param format Audio format (e.g., "m4a")
     * @param messageId Message ID for tracking transcription
     * @return Command configuration with 3 retries and queue enabled
     */
    fun sessionVoiceExecute(
        sessionId: String,
        audioBase64: String,
        format: String,
        messageId: String
    ): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.execute")
            put("id", UUID.randomUUID().toString())
            put("session_id", sessionId)
            put("payload", mapOf(
                "audio" to audioBase64,
                "audio_format" to format,
                "message_id" to messageId
            ))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.execute.voice"
        )
    }

    /**
     * Cancel an agent session execution.
     * @param sessionId ID of the session
     * @return Command configuration with 3 retries and queue enabled
     */
    fun sessionCancel(sessionId: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.cancel")
            put("id", UUID.randomUUID().toString())
            put("session_id", sessionId)
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.cancel"
        )
    }

    // MARK: - Terminal Commands

    /**
     * Send input to a terminal session.
     * @param sessionId ID of the terminal session
     * @param data Input data (text or raw bytes as string)
     * @return Command configuration with 3 retries and queue enabled
     */
    fun terminalInput(sessionId: String, data: String): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.input")
            put("session_id", sessionId)
            put("payload", mapOf("data" to data))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.input"
        )
    }

    /**
     * Resize a terminal session.
     * @param sessionId ID of the terminal session
     * @param cols Number of columns
     * @param rows Number of rows
     * @return Command configuration with 1 retry and no queue (only latest matters)
     */
    fun terminalResize(sessionId: String, cols: Int, rows: Int): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "session.resize")
            put("session_id", sessionId)
            put("payload", mapOf("cols" to cols, "rows" to rows))
        }
        return CommandConfig(
            message = message,
            maxRetries = 1,
            shouldQueue = false,
            debugName = "session.resize"
        )
    }

    /**
     * Request replay of terminal history.
     * @param sessionId ID of the terminal session
     * @param sinceSequence Sequence number to replay from (null for all)
     * @param limit Maximum number of messages to retrieve
     * @return Command configuration with 3 retries and queue enabled
     */
    fun terminalReplay(
        sessionId: String,
        sinceSequence: Long?,
        limit: Int?
    ): CommandConfig {
        val payload = buildMap<String, Any?> {
            if (sinceSequence != null) put("since_sequence", sinceSequence)
            if (limit != null) put("limit", limit)
        }
        val message = buildMap<String, Any?> {
            put("type", "session.replay")
            put("session_id", sessionId)
            if (payload.isNotEmpty()) put("payload", payload)
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "session.replay"
        )
    }

    // MARK: - System Commands

    /**
     * Request state synchronization.
     * @return Command configuration with 3 retries and queue enabled
     */
    fun sync(): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "sync")
            put("id", UUID.randomUUID().toString())
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "sync"
        )
    }

    /**
     * Request audio data for a message.
     * @param messageId The message ID to request audio for
     * @param type Type of audio ("output" or "input")
     * @return Command configuration with 3 retries and queue enabled
     */
    fun audioRequest(messageId: String, type: String = "output"): CommandConfig {
        val message = buildMap<String, Any?> {
            put("type", "audio.request")
            put("id", UUID.randomUUID().toString())
            put("payload", mapOf(
                "message_id" to messageId,
                "type" to type
            ))
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "audio.request"
        )
    }

    /**
     * Request chat history for supervisor or agent session.
     * Protocol v1.13: sync.state no longer includes history - clients must request it explicitly.
     *
     * @param sessionId Target session ID (null for supervisor)
     * @param beforeSequence Load messages BEFORE this sequence (for scroll-up pagination)
     * @param limit Maximum number of messages to return (default: 20, max: 50)
     * @return Command configuration with 3 retries and queue enabled
     */
    fun historyRequest(
        sessionId: String? = null,
        beforeSequence: Long? = null,
        limit: Int? = 20
    ): CommandConfig {
        val payload = buildMap<String, Any?> {
            if (sessionId != null) put("session_id", sessionId)
            if (beforeSequence != null) put("before_sequence", beforeSequence)
            if (limit != null) put("limit", limit)
        }
        val message = buildMap<String, Any?> {
            put("type", "history.request")
            put("id", UUID.randomUUID().toString())
            if (payload.isNotEmpty()) put("payload", payload)
        }
        return CommandConfig(
            message = message,
            maxRetries = 3,
            shouldQueue = true,
            debugName = "history.request"
        )
    }
}
